### PR TITLE
API-4338 - Updating error message to include client_credentials.

### PR DESCRIPTION
**Pull request by bdgrider**

### CODE REVIEW

**Review by lukemarkwordtlibertyits — CHANGES_REQUESTED** [review]
This is functional and is more descriptive than the last error. I am not sure it meets the following criteria `consider rewording to better handle a request with a missing grant_type`. I would recommend either refactoring to check for missing grant types. At the very least adding that the grant type may be missing in the error_description.

File: oauth-proxy/oauthHandlers/tokenHandlerStrategyClasses/tokenStrategies/unsupportedGrantStrategy.js [hunk 1]
```diff
@@ -6,7 +6,7 @@ class UnsupportedGrantStrategy {
       statusCode: 400,
       error: "unsupported_grant_type",
       error_description:
-        "Only authorization and refresh_token grant types are supported",
+        "Only authorization_code, refresh_token, and client_credential grant types are supported",
```
Comment (beaugrantham-va) [inline oauth-proxy/oauthHandlers/tokenHandlerStrategyClasses/tokenStrategies/unsupportedGrantStrategy.js:9]: https://tools.ietf.org/html/rfc6749#section-4.4.2 - `client_credentials`

File: oauth-proxy/oauthHandlers/tokenHandlerStrategyClasses/tokenHandlerClientBuilder.js [hunk 2]
```diff
@@ -167,6 +167,14 @@ const getStrategies = (
       getPatientInfoStrategy: new GetPatientInfoFromLaunchStrategy(req),
     };
   } else {
+    if (req.body.grant_type === undefined || req.body.grant_type === "") {
+      throw {
+        statusCode: 400,
+        error: "empty_grant_type",
```
Comment (beaugrantham-va) [inline oauth-proxy/oauthHandlers/tokenHandlerStrategyClasses/tokenHandlerClientBuilder.js:173]: `empty_grant_type` is not a valid error. I think this falls under `invalid_request`.

- [ ] Update error

File: oauth-proxy/oauthHandlers/tokenHandlerStrategyClasses/tokenHandlerClientBuilder.js [hunk 3]
```diff
@@ -167,6 +167,14 @@ const getStrategies = (
...
+        error_description:
+          "A grant type is required. Supported grant types are authorization, refresh_token, and client_credential.",
```
Comment (beaugrantham-va) [inline oauth-proxy/oauthHandlers/tokenHandlerStrategyClasses/tokenHandlerClientBuilder.js:175]: https://tools.ietf.org/html/rfc6749#section-4.4.2 - `client_credentials`

https://tools.ietf.org/html/rfc6749#section-4.1.3 - `authorization_code` (thanks for updating the other one)

**Review by beaugrantham-va — CHANGES_REQUESTED** [review]

**Comment by beaugrantham-va** [comment]
Please add a unit test and integration test for missing grant type (if not already covered).

- [ ] Unit test for missing grant_type
- [ ] Integration test for missing grant_type

**Review by dbrown-va — APPROVED** [review]

File: oauth-proxy/oauthHandlers/tokenHandlerStrategyClasses/tokenStrategies/unsupportedGrantStrategy.js [hunk 4]
```diff
@@ -4,9 +4,9 @@ class UnsupportedGrantStrategy {
   async getToken() {
     throw {
       statusCode: 400,
-      error: "unsupported_grant_type",
```
Comment (beaugrantham-va) [inline oauth-proxy/oauthHandlers/tokenHandlerStrategyClasses/tokenStrategies/unsupportedGrantStrategy.js:7]: `unsupported_grant_type` was correct here.

- [ ] Change this back

**Review by beaugrantham-va — CHANGES_REQUESTED** [review]

File: oauth-proxy/oauthHandlers/tokenHandlerStrategyClasses/tokenHandlerClientBuilder.js [hunk 5]
```diff
@@ -167,6 +167,14 @@ const getStrategies = (
       getPatientInfoStrategy: new GetPatientInfoFromLaunchStrategy(req),
     };
   } else {
+    if (req.body.grant_type === undefined || req.body.grant_type === "") {
```
Comment (lukemarkwordtlibertyits) [inline oauth-proxy/oauthHandlers/tokenHandlerStrategyClasses/tokenHandlerClientBuilder.js:170]: grant_type could also be null. Truthy check here would cover all the bases 
```
if(!req.body.grant_type)
```

**Review by lukemarkwordtlibertyits — CHANGES_REQUESTED** [review]

**Review by crolarlibertyva — APPROVED** [review]

**Review by dbrown-va — APPROVED** [review]

**Pull request merged**